tokio/src/io/split.rs

```diff
@@ -27,6 +27,10 @@ pub struct WriteHalf<T> {
     inner: Arc<Inner<T>>,
 }
 
+/// An opaque ID for the parent stream of a split half.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub struct SplitStreamId(usize);
```

Comment:
Does it make sense to add a lifetime to `SplitStreamId`, so it can be associated with a `ReadHalf` or `WriteHalf`, and therefore cannot outlive its `Arc<Inner<T>>`? That would prevent having dangling `SplitStreamId`s (Or ABA-issues).

Reply:
Or you could simply drop `SplitStreamId` and directly implement `std::cmp::PartialEq<...>` on `ReadHalf`/`WriteHalf`.

Reply:
I see your point. A liftetime would make it impossible to use the ID as a key in an associative container where the value contains the half.

Implementing `PartialEq`, `PartialOrd` and `Hash` directly on the halves should probably work without the danger of dangling `SplitStreamId`. I'd like to try it out a bit though, to make sure there aren't any annoying practical problems.

Reply:
Since neither `ReadHalf` nor `WriteHalf` can be `Borrow` of the other, looking up one in a map or set given the other would be hard without a separate `SplitStreamId` type.

Reply:
Right, I forgot about the borrow bound for looking up things in a map/set. That doesn't really leave another option besides a `SplitStreamId` that can dangle.

Perhaps the best thing to do is simply to document this on `SplitStreamId` type and the `stream_id` functions and move the responsibility of not comparing dangling IDs to the user? This shouldn't be hard in practice.

Reply:
It's not a big deal if they 'dangle' anyway. The `unsplit` function will check it again when used, which will guarantee both halves are alive.

Reply:
I added some documentation to explain that a `SplitStreamId` can dangle.